R/trackhub.R

```diff
@@ -376,58 +404,10 @@ readAndSanitize <- function(filepath) {
     contentDf
 }
 
-parseFileAndAddTracks <- function(tabCountList, contentDf) {
-    trackList <- list()
-    childrenList <- list()
-    parentLookUp <- list()
-    tracksIndex <- grep("track", contentDf$V1)
-    tracksIndex <- vapply(tracksIndex, function(x) x-1, numeric(1))
-    totalLines <- length(contentDf$V1)
-    tracksIndex[[length(tracksIndex) + 1]] <- totalLines
-    totaltracks <- length(tracksIndex)
-    start <- 1L
-    end <- 2L
-    parentTrackName <- ""
-    trackListIndex <- 1L
-
-    while (start != totaltracks) {
-        previousLevel <- tabCountList[tracksIndex[end - 1]]
-        level <- tabCountList[tracksIndex[end]]
-        track <- setNames(
-            as.list(contentDf$V2[(tracksIndex[start]+1):tracksIndex[end]]),
-            contentDf$V1[(tracksIndex[start]+1):tracksIndex[end]]
-        )
-        if (level == -1L) {
-            trackList[[trackListIndex]] <- track
-            parentTrackName <- contentDf$V2[(tracksIndex[start]+1)]
-            trackListIndex <- trackListIndex + 1
-        }else{
-            if (level != previousLevel)
-                parentTrackName <- contentDf$V2[(tracksIndex[start-1]+1)]
-            parentLookUp[[length(parentLookUp) + 1]] <-
-                c(
-                  paste0(parentTrackName,"/",contentDf$V2[(tracksIndex[start]+1)])
-                )
-            if (is.null(childrenList[[parentTrackName]]))
-                childrenList[[parentTrackName]] <- list()
-            childrenIndex <- length(childrenList[[parentTrackName]]) + 1
-            childrenList[[parentTrackName]][[childrenIndex]] <- track
-        }
-        start <- start + 1
-        end <- end + 1
-    }
-    return(list(trackList, childrenList, parentLookUp))
-}
-
-TrackContainer <- function(filepath) {
-    contentDf <- readAndSanitize(filepath)
-    tabCountList <- getTabCountList(contentDf)
-    contentDf$V1 <- gsub("^(\\t)+", "", contentDf$V1)
-    container <- parseFileAndAddTracks(tabCountList, contentDf)
-    tc <- new("TrackContainer",
-              trackList = container[[1]],
-              childrenList = container[[2]],
-              parentLookUp = container[[3]]
-    )
-    tc
+getTabCountList <- function(contentdf) {
+    tabCountList <- vapply(contentdf$V1, function(x) {
```

Comment:
The `gregexpr()` call is vectorized so does not need to be in the loop. It is only necessary to loop to retrieve the `"match.length"` attribute. Could be done pretty efficiently using `vapply(matches, attr, integer(1L), "match.length")`. It's integer, not numeric, btw.